Vue.js activity template

dhruvmisra opened the pull request:
# Vue.js Activity Template #
This is an Activity Template to create activities in Vue.js. Currently it is an activity named 'ActivityTemplateVue' and runs the basic Pawn activity for testing purposes.

## Structure ##
I have decomposed the major functionalities into importable components which can be found under `/js/components`.

Steps to use any component (Adding the toolbar for example):
1. Import the component using `<script>` tag in `index.html`
```html
  ...
  <script src="js/components/SugarToolbar.js"></script>    //Import the toolbar component
  <script src="js/activity.js"></script>              //Make sure you import activity.js in the end
</body>
```
2. Create and use the component instance in `index.html` (Don't forget the `ref` attribute)
```html
<div id="app">
  <sugar-toolbar ref="SugarToolbar">
    <sugar-toolitem id="add-button" v-on:click="onAddClick" v-bind:title="l10n.stringAddPawn"></sugar-toolitem>
  </sugar-toolbar>
  ...
```

## Benefits ##
* Easy to integrate a specific functionality.
* Decomposed and encapsulated code making it easier to manage. 
* More flexibility in project design.
* Adds a layer of abstraction. New contributors can focus on activity-specific code rather than the working of Sugarizer functionalities.

## Can be improved ##
* Palette workflow
* ~Localization workflow~

This is an initial template structure and is open to changes. All suggestions are welcome.

llaske commented:
That's not bad.
But it's not exactly what I expected.
In my idea what we call "component" is something generic where no code specific to the activity should be include. Here there is code specific to the Pawn activity in the component files.

More detailed remarks:
- What you call "Journal" is not really the Journal but a mix between activity and initialization
- sendMessage could be a method of the Presence component
- Why not create Vue event on the Presence component to match presence event
- I don't think we need a presence object in activity data since the Presence component should handle everything

Note also that you've mixed different indentation styles in files (TAB/space) and that there is some ES6 syntax (let).

dhruvmisra commented:
I have the components more generic and handled all the activity specific code outside. 
* Created callback functions and methods to handle data outside the component.
* Emitting Vue events from Presence to handle DataReceived and UserChanged.
* The presence object in activity data is now an instance of presence Vue component, rather than the utility itself.

llaske commented on activities/ActivityTemplateVue.activity/js/activity.js:
```diff
@@ -0,0 +1,107 @@
+		},
+		full: false,
+		context: null,
+		presence: null,
```
I think it's not need

dhruvmisra replied:
Workflow has been changed.

llaske commented on activities/ActivityTemplateVue.activity/js/activity.js:
```diff
@@ -0,0 +1,107 @@
+		var vm = this;
+		requirejs(["sugar-web/activity/activity", "sugar-web/env", "sugar-web/graphics/icon"], function (activity, env, icon) {
+			// Initialize Sugarizer
+			activity.setup();
```
Why not rename your Journal component to an Activity component then implement this in a setup method of the Activity component?

dhruvmisra replied:
Hmm, interesting. I'll start testing this workflow.

llaske commented on activities/ActivityTemplateVue.activity/js/activity.js:
```diff
@@ -0,0 +1,107 @@
+			// Initialize Sugarizer
+			activity.setup();
+			
+			env.getEnvironment(function (err, environment) {
```
May be getEnvironment could be a sync method of the Activity component. And the activity could handle by itself the loading of the environment at mount

dhruvmisra replied:
Made the changes

llaske commented on activities/ActivityTemplateVue.activity/js/activity.js:
```diff
@@ -0,0 +1,107 @@
+				vm.currentenv = environment;
+			});
+
+			vm.icon = icon;
```
Here again you could do a component instead

dhruvmisra replied:
Created a Pawn component which uses the icon library in a8302fe.

llaske commented on activities/ActivityTemplateVue.activity/js/components/journal.js:
```diff
@@ -0,0 +1,54 @@
+var Journal = {
+  data: {
```
Mixed indentation style in this file

dhruvmisra replied:
Should be fixed now.

llaske commented on activities/ActivityTemplateVue.activity/js/components/journal.js:
```diff
@@ -0,0 +1,54 @@
+					activity.getDatastoreObject().loadAsText(function (error, metadata, data) {
+						if (error == null && data != null) {
+							let context = JSON.parse(data);
+							vm.parent.pawns = context.pawns;
```
This code is specific it should not be there.

dhruvmisra replied:
Components are more generic now.

llaske commented on activities/ActivityTemplateVue.activity/js/components/journal.js:
```diff
@@ -0,0 +1,54 @@
+				// Load context
+				if (environment.objectId) {
+					console.log("Existing instance");
+					activity.getDatastoreObject().loadAsText(function (error, metadata, data) {
```
I wonder if all callback should not be implemented as event

dhruvmisra replied:
Emits an event on data load now.

llaske commented on activities/ActivityTemplateVue.activity/js/components/journal.js:
```diff
@@ -0,0 +1,54 @@
+		parent: null,
+	},
+  mounted() {
+		this.parent = this.$parent;
```
Think it's not need here

dhruvmisra replied:
Removed this property.

llaske commented on activities/ActivityTemplateVue.activity/js/components/presence.js:
```diff
@@ -0,0 +1,101 @@
+				return;  	// Return if data was sent by the user
+			}
+
+			switch (msg.content.action) {
```
This code is specific if should not be in the component

dhruvmisra replied:
Components are more generic now.

llaske commented on activities/ActivityTemplateVue.activity/js/components/presence.js:
```diff
@@ -0,0 +1,101 @@
+        
+			});
+
+			vm.humane = humane;
```
I think we could remove humane stuff here. May be we could add a Popup component to handle all humane features?

dhruvmisra replied:
Yes, we can do that.

dhruvmisra replied:
Created the Popup component

llaske commented on activities/ActivityTemplateVue.activity/js/components/presence.js:
```diff
@@ -0,0 +1,101 @@
+  },
+  methods: {
+    onNetworkDataReceived: function (msg) {
+			if (this.presence.getUserInfo().networkId === msg.user.networkId) {
```
Mixed indentation in this file

dhruvmisra replied:
Should be fixed now.

llaske commented on activities/ActivityTemplateVue.activity/js/components/toolbar.js:
```diff
@@ -0,0 +1,73 @@
+}
+
+// Toolbar component
+var Toolbar = {
```
I wonder if we need a Toolbar component. Only ToolbarItem is a generic component. May be we could be directly all Toolbar code in the main activity file.

dhruvmisra replied:
It would be a good practice to have the ToolbarItems in their correct container. The ToolbarItems would give us finer control over the buttons and the Toolbar will help us define behaviour of the toolbar. I think the fullscreen methods should be moved to Toolbar component.

llaske commented on activities/ActivityTemplateVue.activity/js/palettes/ballpalette.html:
```diff
@@ -0,0 +1,34 @@
+<div class="ball-select">
```
Thinks that you could remove this palette and all related images

dhruvmisra replied:
Yes, I'll do that. Will add a palette component later.

llaske commented:
My feedbacks:
- I think the version without Event Bus is better. I'm not sure we could have a very complex structure of components in an activity. So I'm not sure is useful.
- As I said, we don't need to reference the activity object, in the Journal object, we could get the sugar-web activity object by a require call.

llaske commented:
That's better.

I'm going to look more but here's my first remarks:
- The Toolbar items are defined both in the `toolbar.js` component and in the `index.html`. Items in the toolbar should be defined at one place only. I think it's better if everything is in `index.html`. 
- ToolbarItem are able to handle automatically linked palette (see for example [here](https://github.com/llaske/sugarizer/blob/dev/activities/Calligra.activity/js/toolbar.js#L61)). You don't need to write code for it in `presence.js`
- The test `this.presence.presence` [here](https://github.com/llaske/sugarizer/pull/800/files#diff-2d321821ee927c03050db8743e13b1efR64) should be done by a call to a `isConnected()` method.

dhruvmisra commented:
> The Toolbar items are defined both in the `toolbar.js` component and in the `index.html`.

Only the buttons which are essential for each activity i.e. Activity, Help, Fullscreen and Stop buttons are present inside the Toolbar component. Also the network button is present which is conditionally rendered automatically if the Presence component is being used. I think it makes for a cleaner `index.html` which should only contain buttons specific to the activity.

>ToolbarItem are able to handle automatically linked palette

Yes you're right, let me try to handle presence palette in ToolarItem.

>The test this.presence.presence here should be done by a call to a isConnected() method.

Yes this will be better. Adding this next.

llaske commented on activities/ActivityTemplateVue.activity/js/components/toolbar.js:
```diff
@@ -32,6 +32,16 @@ Vue.component('toolbar-item', {
 			});
 		}
+
+		if(vm.id == 'network-button') {
+			requirejs(["sugar-web/graphics/presencepalette"], function(presencepalette) {
+				var presencePalette = new presencepalette.PresencePalette(document.getElementById(vm.id), undefined);
+				presencePalette.addEventListener('shared', function () {
+					presencePalette.popDown();
+					vm.$root.$refs.presence.onShared();
+				});
+			});
```
You don't need to specify the source code. Just use the `paletteFile`, `paletteClass` and `paletteEvent` in `toolbar-item`. See for example [here](https://github.com/llaske/sugarizer/blob/dev/activities/Calligra.activity/js/toolbar.js#L51).

dhruvmisra replied:
The attributes `paletteFile`, `'paletteClass`, etc are used to create a normal Palette. The presence palette has a different library which needs to be called I believe. Can the presence palette be created with `new palette` instead of `new presencepalette`?

llaske replied:
It's a bit tricky but the `paletteClass` is use to create the object with the right JavaScript class. See the source code of this call [here](https://github.com/llaske/sugarizer/blob/dev/activities/Calligra.activity/js/toolbar.js#L32).
So it will works if you're using `'presencepalette'` value for `paletteClass` property.
In a same way if you're using `'shared'` value for `paletteEvent` property, you will be able to use `v-on:shared` syntax to handle the event.

dhruvmisra replied:
Wow that's clever! I'll test it soon.

dhruvmisra replied:
Added in c428dd8

llaske commented on activities/ActivityTemplateVue.activity/js/activity.js:
```diff
@@ -13,57 +13,50 @@ var app = new Vue({
+			function initialLocalization() {
+				vm.displayText = vm.SugarLocalization.l10n.get("Hello", { name: vm.currentenv.user.name });
+				vm.SugarLocalization.localize(vm.l10n);
+				window.removeEventListener('localized', initialLocalization);
```
Not sure why you need to do that

dhruvmisra replied:
I created this to show the welcome message only once during the activity, and not be affected by any localized event. But I think the event is fired only once during the runtime of any activity. I'll add the message code to the localized method.

dhruvmisra replied:
Removed the `initialLocalization()` and made flow better in 06d7134.

dhruvmisra commented:
> Can you imagine to do it more generic?

Yes I'm currently working to make it more generic. We can send the steps of the tutorial as a prop and extract the `l10n` strings from these steps and handle localization. I had a doubt regarding other methods, Bootstrap Tour API provides a number of functions like `onStart()`, `onEnd()`, `onShow()`, etc., should there be a method for each one of these? As they might be useful in some cases.


> I'm not a big fan of the use of `$nextTick` method. Is there any way to do that?

`$nextTick` function is the recommended way to perform some action right after Vue has completed updating the DOM. (Usually after changing data variables, [see here](https://vuejs.org/v2/api/#vm-nextTick)). An alternative would be to add a small timeout to wait for updation but that would be a bad approach in my opinion.

llaske commented:
Regarding Tutorial, I don't think we need to encapsulate all bootstrap methods. I think that giving a way to set steps is enough.

Regarding `$nextTick`, I think a better way to do that in Vue.js should be to create a `Pawn` component, handle color as a property of this component then use `v-for` to display all pawns. In that case I think colorize method should not be in the `SugarActivity` component but just call from `requirejs` in the `Pawn` component itselft.

dhruvmisra commented:
Yes good idea, we can have the Pawn component get the colors as props and set the color in `mounted()`.

As for the Tutorial, I have added basic event emits for these: `onStart()`, `onShow()`, `onEnd()`. I saw these were used in some activities.

dhruvmisra commented on activities/ActivityTemplateVue.activity/js/activity.js:
```diff
@@ -0,0 +1,107 @@
+			this.pawns.push(this.currentenv.user.colorvalue);
+			this.drawPawns();
+
+			if (this.presence) {
```
Added `isConnected()` to test for presence existence.

llaske commented:
Good!
I specifically do like your `Pawn` component.

I wonder if we really need a `loaded` event in `SugarActivity`. May be we could remove it and call directly the sugar-web `setup` method in the `mounted` event, so the user don't need to call it (and we could remove the method from the component). 

Minor updates:
- All files should respect the same indentation style (TAB)
- What you call ActivityTemplateVue.activity is the Pawn.activity implemented in Vue.js. In fact, at end you should have two directories: one for Pawn.activity and one for ActivityTemplateVue

dhruvmisra commented:
Separated the Activity Template from the Pawn activity. Also made some changes to the workflow as per our discussion.
* Moved `activity.setup()` to SugarActivity component
* Removed SugarTutorial's dependency on SugarL10n. Localized strings should be passed in steps now.
* Reworked the insertFromJournal filters; Added support for all type of objects.
* Refactored the code and removed some useless methods.

llaske commented:
It's perfect. Nice job.

Few (minor) remarks:
- `ActivityTemplateVue` should not be a part of `activities.json` (it's not an activity)
- It will be better to have only on ActivityTemplate directory with two subdirectories `VanillaJS` and `VueJS` (will require to change the text of the initial tutorial to mention that)
- `po` directory should be remove from the activity and from the template
- `README.md` file should be remove from the activity and from the template
- Please update `lib/sugar-web/presencepalette.js` with the latest version in ActivityTemplate in dev branch (this new version automatically handle add/remove users in the palette)
- Please update `LICENSE` file with the latest version in `ActivityTemplate` in dev branch
- Please add `NOTICE` file with the one in `ActivityTemplate` in dev branch
- Update same indentation style (TAB) in all files

dhruvmisra commented:
Made these changes, updated the tutorial text to reflect the new folder structure.

llaske commented:
Nice.

Few remarks:
- The default background in the ActivityTemplate should be gray else the background in the tutorial is not the same (white instead of gray on current screen capture)
- For the same reason, the fullscreen button in the toolbar is not in the VanillaJS tutorial, It should not be in the VueJS version
- There is an error in the console when the Pawn activity is shared:

```
TypeError: Cannot read property 'id' of undefined
    at a.sendMessage (SugarPresence.js:44)
    at wn.onAddClick (activity.js:60)
    at He (vue.min.js:6)
    at a.n (vue.min.js:6)
    at He (vue.min.js:6)
    at a.e.$emit (vue.min.js:6)
    at HTMLDivElement.<anonymous> (SugarToolbar.js:31)
    at HTMLDivElement.<anonymous> (pawnpalette.js:18)
```

dhruvmisra commented:
Fixed the background color and removed the fullscreen buttons. 
I wasn't able to replicate this issue on my side, sharing the activity doesn't show console any error for me. Can you tell me the steps you followed @llaske?

llaske commented:
Hmmm. Look like it was related to my browser cache.
BTW the background is still white on my side.

dhruvmisra commented:
Should be fixed now.

llaske commented:
Thanks.
Not sure where you've updated tutorial text. I've updated `tutorial_step1.md` to mention the new `VanillaJS` directory.
